aten/src/ATen/native/Distance.cpp

```diff
@@ -75,4 +88,269 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
+    grad.size(0), " != size(1) ", grad.size(1));
+  int64_t d = grad.size(0);
+  int64_t size = d * (d - 1) / 2;
+  Tensor out_tensor = at::zeros({size}, grad.options());
+  AT_DISPATCH_ALL_TYPES(self.type(), "squareform", [&] {
+      to_squareform_backward_frame<scalar_t> (grad.data<scalar_t>(),
+        out_tensor.data<scalar_t>(), d, size);
+      }
+      );
+  return out_tensor;
+}
+
+Tensor from_squareform_backward_cpu(const Tensor& grad, const Tensor& self)
+{
+  AT_CHECK(grad.dim() == 1,
+    "Expected 1D grad, but the dimension of the grad is ", grad.dim());
+  int64_t size = grad.size(0);
+  int64_t d = getSquareFormDimension(size);
+  Tensor out_tensor = at::zeros({d, d}, grad.options());
+  AT_DISPATCH_ALL_TYPES(self.type(), "squareform", [&] {
+      from_squareform_backward_frame<scalar_t> (grad.data<scalar_t>(),
+        out_tensor.data<scalar_t>(), d, size);
+      }
+      );
+  return out_tensor;
+}
+
+Tensor squareform_cpu(const Tensor& self)
+{
+  AT_CHECK((self.dim() == 1 || self.dim() == 2),
+    "expected vector-form distance or square-form distance as input",
+    " but the dimension got from input is ", self.dim());
+  if (self.dim() == 1)
+  {
+    return to_squareform_cpu(self);
+  }
+  else
+  {
+    return from_squareform_cpu(self);
+  }
+}
+
+Tensor squareform_backward_cpu(const Tensor& grad, const Tensor& self)
+{
+  AT_CHECK((grad.dim() == 1 || grad.dim() == 2),
+    "expected the input dim is 1 or 2, but got ", grad.dim());
+  if (grad.dim() == 1)
+  {
+    return from_squareform_backward_cpu(grad, self);
+  }
+  else
+  {
+    return to_squareform_backward_cpu(grad, self);
+  }
+}
+
+Tensor to_squareform_cuda(const Tensor& self)
+{
+  AT_CHECK(self.dim() == 1,
+    "expected vector-form distance as input, dim as 1, but got ", self.dim());
+  int64_t size = self.size(0), d = (std::sqrt(8 * size + 1) + 1) / 2;
+  AT_CHECK(d * (d - 1) / 2 == size,
+    "not able to derive dimension, d from size ", size);
+  Tensor out_tensor = at::zeros({d, d}, self.options());
+  for (int64_t i = 0, idx = 0; i < d; i++)
+  {
+    for (int64_t j = i + 1; j < d && idx < size; j++)
+    {
+      out_tensor[i][j] = self[idx];
+      out_tensor[j][i] = self[idx];
+      idx++;
+    }
+  }
+  return out_tensor;
+}
+
+Tensor from_squareform_cuda(const Tensor& self)
+{
+  AT_CHECK(self.dim() == 2,
+    "expected square-form distance as input, dim as 2, but got ", self.dim());
+  AT_CHECK(self.size(0) == self.size(1),
+    "expected square-form distance as input, size(0) ",
+    self.size(0), " != size(1) ", self.size(1));
+  int64_t d = self.size(0), size = d * (d - 1) / 2;
+  Tensor out_tensor = at::zeros({size}, self.options());
+  for (int64_t i = 0, idx = 0; i < d; i++)
+  {
+    for (int64_t j = i + 1; j < d && idx < size; j++)
+    {
+      out_tensor[idx] = self[i][j];
+      idx++;
+    }
+  }
+  return out_tensor;
+}
+
+Tensor to_squareform_backward_cuda(const Tensor& grad, const Tensor& self)
+{
+  AT_CHECK(grad.dim() == 2,
```

Comment:
this cuda stuff shouldn't be here, this isn't a cuda implementation.  just check if the input is cuda and throw an error (or put in a dispatch only for CPU).